### PR TITLE
feat: Rhesis Architect — AI agent for test suite design and execution

### PR DESCRIPTION
**Pull request by harry-rhesis**
## Purpose

Introduces the **Rhesis Architect**, a conversational AI agent that guides users through designing, generating, and running test suites for AI applications. The agent combines an MCP-based tool layer with a streaming chat UI, enabling end-to-end test suite creation from a single conversation.

### Agent Framework (`sdk/`)
- Add `agents/` framework with a base agent loop, MCP tool client, and structured JSON reasoning format
- Add `LocalToolProvider` for in-process tool dispatch inside the Celery worker (no MCP round-trip needed)
- Add `ExploreEndpointTool` for multi-turn endpoint capability discovery via Penelope
- Add exploration strategies: `domain_probing`, `capability_mapping`, `boundary_discovery`, and `comprehensive`
- Add `ArchitectAgent` with full phase-based workflow: Discovery → Planning → Creating → Execution
- Add plan tracking, write-guard confirmation flow, and auto-approve mode
- Add `await_task` tool for signal-based async job notification (replaces manual polling)
- Add `metric synthesizer` for AI-generated metrics with structured evaluation criteria
- Add knowledge source grounding: pass source IDs to `generate_test_set` for context-aware test generation
- Add result analysis helpers: run comparison, per-behavior/metric stats, failure drill-down
- Fix litellm aiohttp transport to prevent event-loop errors in Celery worker threads

### MCP Server (`apps/backend/src/rhesis/backend/app/mcp_server/`)
- Add full MCP server mounted at `/mcp`, with auth middleware and tool dispatch via ASGI transport
- Split monolithic `mcp_server.py` into a package (`tools.py`, `server.py`, `local_tools.py`, `schema.py`)
- Add `mcp_tools.yaml` — single source of truth for tool names, descriptions, HTTP mappings, and pagination config
- Add `page_size` field to all list tools: server enforces peek-ahead pagination and wraps responses in a `_pagination` envelope so the LLM always knows whether more results exist
- Remove `limit` from LLM-visible schemas for paginated tools so the agent cannot override server-managed page sizes
- Add `apply_query_overrides` and `format_list_response` helpers shared between `server.py` and `local_tools.py`
- Add `list_sources` tool for knowledge source discovery
- Add `explore_endpoint`, `save_plan`, `await_task`, `get_test_result_stats`, `get_test_run_stats`, `improve_metric`, `generate_metric`, and run comparison tools

### Backend (`apps/backend/`)
- Add `architect_chat_task` Celery task: runs the agent, streams events over WebSocket, handles tool confirmation
- Add `signal-based async task notification` so `await_task` resumes the agent when a background job completes instead of polling
- Add delegation token auth so the Celery worker can call backend routes as the originating user
- Add `$select` support to sources and tests endpoints (previously missing, causing context overflow when listing large collections)
- Add `$select` to all remaining list endpoints that were missing it (behaviors, categories, topics, metrics, projects, test sets, test runs, test results)
- Add `check_endpoint` tool for connectivity verification before exploration
- Fix MCP server lifespan, session manager lifecycle, and auth consolidation

### Frontend (`apps/frontend/`)
- Add Architect chat page with full streaming UI: reasoning steps, tool call bubbles with labels, confirmation buttons, plan display, and result markdown
- Add `@-mention` system: typing `@` shows a live-search dropdown for endpoints, sources, test sets, behaviors, and metrics; resolved entities are injected as structured context
- Add session state management and plan persistence across messages
- Add Telemachus (Architect) personality, logo, and beta nav label
- Fix streaming message backfill, broken entity links, sidebar toggle, and welcome screen race condition
- Fix `@-mention` dropdown to show results on empty query and strip hint markup on selection

### System Prompt (`sdk/.../system_prompt.j2`)
- Full phased workflow: Discovery, Planning, Creating, Execution
- Reuse-before-create discipline for behaviors and metrics
- `$select` guidance for all list calls to keep responses small
- Pagination section explaining `_pagination` envelope, `has_more`, and `next_skip`
- Knowledge source workflow, result analysis patterns, comparison workflow
- Prompt injection resistance, identity boundaries, tool safety rules
- Confirm-before-act policy for all write operations

### Tests
- Add backend architect agent tests
- Add SDK agent tests: targets, MCP agent, metric synthesizer

## Additional Context

- This is the Telemachus project (internal name for the Architect agent initiative)
- The agent runs inside the Celery worker via `LocalToolProvider`, bypassing the MCP protocol for in-process calls while remaining protocol-compatible with external MCP clients
- All list tools use server-managed pagination — the LLM cannot override page sizes, preventing accidental context overflow from large collections

## Testing

1. Start the stack: `./rh dev backend`, `./rh dev worker`, `./rh dev frontend`
2. Navigate to the Architect page
3. Type `@` to verify the mention dropdown shows endpoints and sources
4. Ask "test my [endpoint name]" to run through the full Discovery → Planning → Creating → Execution flow
5. After a test run completes, ask "analyze the results" to verify result analysis tools
6. Ask "what are my sources?" to verify all sources are returned (previously truncated to 2)

image

image

### CODE REVIEW

File: apps/backend/src/rhesis/backend/app/mcp_server/server.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    )
+
+    # ASGI wrapper that validates auth before delegating to the
+    # session manager.  MCP uses raw ASGI (not FastAPI routes), so
+    # AuthenticatedAPIRoute doesn't apply — we check the bearer
+    # token ourselves.
+    class _MCPApp:
+        async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+            if scope["type"] == "http":
+                request = Request(scope, receive, send)
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/mcp_server/server.py:200]: Good catch on the pattern — confirmed that `Request(scope, receive, send)` is valid in Starlette 0.50+: the constructor signature is `Request(scope, receive=empty_receive, send=empty_send)` (three positional args). No change needed here.

File: apps/backend/src/rhesis/backend/app/routers/behavior.py [hunk 2]
```diff
@@ -82,6 +90,10 @@ def read_behaviors(
         organization_id=organization_id,
         user_id=user_id,
     )
+    if select:
+        serialized = jsonable_encoder(results)
+        return JSONResponse(content=apply_select(serialized, select))
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/routers/behavior.py:95]: Fixed. Updated `with_count_header` to copy headers from the injected `response` dependency onto any `Response` subclass returned by the route. This way `X-Total-Count` (and any future headers set by the decorator) are never silently dropped on `$select` paths.

File: apps/backend/src/rhesis/backend/app/mcp_server/server.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/mcp_server/server.py:200]: Verified: Starlette 0.50's `Request.__init__` signature is `(scope, receive=empty_receive, send=empty_send)` — three positional args is valid. No change needed.

File: apps/backend/src/rhesis/backend/app/services/websocket/handlers/architect.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    4. Return acknowledgement to client
+
+    Streaming events arrive via Redis pub/sub on channel
+    ``architect:{session_id}`` and are forwarded by the WS manager
+    to clients subscribed to that channel.
+    """
+    correlation_id = message.correlation_id
+    payload = message.payload or {}
+
+    session_id = payload.get("session_id")
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/services/websocket/handlers/architect.py:45]: Fixed. `handle_architect_message` now calls `crud.get_architect_session()` with the current user's `organization_id` and `user_id` before touching any data. If the session doesn't exist under that tenant, the handler returns an error immediately — no message is persisted.

File: apps/backend/src/rhesis/backend/app/auth/token_utils.py [hunk 4]
```diff
@@ -20,7 +20,7 @@
 logger = logging.getLogger(__name__)
 
 # Token expiry defaults (in minutes)
-SERVICE_DELEGATION_EXPIRE_MINUTES = 5  # 5 minutes, short-lived for service-to-service calls
+SERVICE_DELEGATION_EXPIRE_MINUTES = int(os.getenv("SERVICE_DELEGATION_EXPIRE_MINUTES", "60"))
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/auth/token_utils.py]: Fixed. Default changed to 15 minutes — long enough for typical test-generation jobs with polyphemus, short enough to keep the privilege window reasonable. Still overridable via `SERVICE_DELEGATION_EXPIRE_MINUTES` for deployments that need more.

File: apps/backend/src/rhesis/backend/tasks/architect_monitor.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+_KEY_TTL = 7200  # 2 hours
+
+_redis_pool: Optional[_redis_lib.ConnectionPool] = None
+
+
+def _get_redis() -> _redis_lib.Redis:
+    """Return a Redis client reusing a shared connection pool."""
+    global _redis_pool
+    if _redis_pool is None:
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/tasks/architect_monitor.py:39]: Fixed. `architect_monitor._get_redis()` now uses `BROKER_URL or REDIS_URL` fallback, matching the pattern already used in `redis_pubsub.py` and `publisher.py`.

File: apps/backend/src/rhesis/backend/app/mcp_server/server.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                        scheme="Bearer",
+                        credentials=auth_header[7:],
+                    )
+
+                try:
+                    secret_key = get_secret_key()
+                except Exception:
+                    secret_key = None
+
+                user = await get_authenticated_user_with_context(
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/mcp_server/server.py:217]: Fixed. The `except Exception` now logs at `ERROR` level with the exception message before falling back to `secret_key=None`, so a misconfigured JWT secret surfaces loudly in logs rather than silently making `/mcp` behave as unauthenticated.

File: apps/backend/src/rhesis/backend/app/utils/odata.py [hunk 7]
```diff
@@ -15,6 +20,41 @@ def apply_odata_filter(query: Query, model: Type, filter_expr: str | None) -> Qu
     return query
 
 
+def apply_select(
+    data: Union[List[Dict[str, Any]], Dict[str, Any], Any],
```
[the remark this answers is not in the document]
Comment (harry-rhesis) [reply apps/backend/src/rhesis/backend/app/utils/odata.py:24]: Fixed. The `__dict__` fallback in `_pick` now filters out keys starting with `_`, which covers `_sa_instance_state` and any other private/internal attrs. In practice all callers already pass `jsonable_encoder()` output (plain dicts), so this is a safety net for future misuse.
